src/finite_gp_projection.jl

```diff
@@ -23,6 +23,17 @@ end
 function FiniteGP(
     f::AbstractGP, X::AbstractMatrix, σ²=default_σ²; obsdim::Int=KernelFunctions.defaultobs
 )
+    nrows, ncols = size(X)
+
+    if obsdim == 1
+        ndata, dim, xvecs = nrows, ncols, "RowVecs"
+    elseif obsdim == 2
+        ndata, dim, xvecs = ncols, nrows, "ColVecs"
+    else
+        throw(ArgumentError("obsdim must be 1 or 2"))
+    end
+    @info "Evaluating the GP on a $nrows × $ncols matrix, interpreted as $ndata points of dimension $dim. To remove this message, wrap the matrix in $xvecs."
```

Comment:
I think we should either use `@debug`, show the message only in the case of an error, or properly deprecate support for matrices (would be inconsistent with KernelFunctions though). Otherwise we mainly annoy users with logs.

Reply:
Showing only in case of error lets it work (but wrongly) when X is square (though that's probably a rare case). If any user doesn't want to see the message anymore, they can simply get rid of it by wrapping in ColVecs as suggested, no? We could be firmer on "we recommend using ColVecs(X) instead"...

Reply:
My point is: either we support it or we don't. But if we support it we should beither spam the users no cause slowdowns (logging is quite slow).

Reply:
I think it's nice to have it give a meaningful message, but still continue running. I believe that will make the transition easier for users coming from other packages. (Instead of outright deprecating passing a matrix, and eventually dropping support entirely.)

How about
```suggestion
    @info "For vector-of-vectors input represented by a matrix X, please use f(ColVecs(X)) or f(RowVecs(X)) instead (depending on your data layout). f(X) is now evaluating the GP on a $nrows × $ncols matrix, interpreted as $ndata points of dimension $dim."
```
?

Reply:
No, as a user I don't want to see messages if it's supported. There's not something like supporting it a bit, either it's part of the API or not.

If you want users to move to ColVecs etc then deprecating it is the right approach. We can still support it for some longer time, it doesn't mean we have to remove it in the next breaking release.

I'm fine with deprecating it if we agree on this and are consistent throughout different packages (thus eg requires changes in KernelFunctions as well).

Reply:
If you don't want to see messages, just use `f(ColVecs(X))`. What's the problem with offering people a way of doing it how they might naively expect it to work by just passing in a matrix, and getting an explanation of what's happening and what they should do instead?

Reply:
`@info` is not designed for this use case. This is exactly what deprecations are for.

Reply:
I disagree with that. The point here is to take something that a new user might expect to do (call `f(X)` with a matrix as they might expect from python or matlab GP codes), and tell them what they should be doing instead. `@deprecate` doesn't allow for any custom messages.

I'd like to hear what everyone else is thinking on this. @willtebbutt , @theogf ?

Reply:
Another issue with deprecations is that they won't print a message by default, and I wouldn't expect a new user to think of running `julia --depwarn=yes`. The point of this _is_ that by default you do get a message. I see it as a "make it easy for users to learn".

If we want to change it to always require obsdim, we'll have to change the APIs - in the current API you can't pass `obsdim` - `FiniteGP` itself isn't exported, the public API expects you to just call `f(X)`.

Reply:
This discussion doesn't seem to lead anywhere unfortunately :shrug: I've said everything I want to say already above.

> If we want to change it to always require obsdim, we'll have to change the APIs - in the current API you can't pass `obsdim` - `FiniteGP` itself isn't exported, the public API expects you to just call `f(X)`.

No, we wouldn't have to change the API. We would just add a dispatch on `X` in https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/blob/e7c67bd24c454f0b8179032ecbb82284887159c0/src/finite_gp_projection.jl#L38 and wrap matrices in `ColVecs` or `RowVecs` already at this stage (and probably deprecate the non-exported `FiniteGP` constructor with `X::AbstractMatrix`).

Reply:
We discussed this at the meeting today, and it was decided to make it error explicitly, so I'll resolve this thread now.